File: wasm/src/runtime/filtered-plane.ts
import type {
  ToolCall, ToolSchema, StreamEvent, ToolResultEvent, ToolDeniedEvent,
} from "../types.js"
import type { RegisteredTool } from "../tools/index.js"
import type { ExecutionPlane, RunContext } from "./execution-plane.js"

const DEFAULT_META_TOOLS = new Set(["skill", "memory", "knowledge", "update_plan"])

/** Wraps an execution plane, allowing only manifest-permitted tool IDs (+ meta-tools). */
export class FilteredExecutionPlane implements ExecutionPlane {
  constructor(
    private readonly inner: ExecutionPlane,
    private readonly permittedIds: Set<string>,
    private readonly metaTools: Set<string> = DEFAULT_META_TOOLS,
  ) {}

  register(...tools: RegisteredTool[]): this {
    this.inner.register(...tools)
    return this
  }

  unregister(name: string): this {
    this.inner.unregister(name)
    return this
  }

  schemas(): ToolSchema[] {
    return this.inner.schemas().filter(s => this.permittedIds.has(s.name) || this.metaTools.has(s.name))
  }

  async *executeAll(calls: ToolCall[], ctx: RunContext): AsyncIterable<StreamEvent> {
    const permitted: ToolCall[] = []
    for (const call of calls) {
      if (this.metaTools.has(call.name) || this.permittedIds.has(call.name)) {
        permitted.push(call)
        continue
      }
      const reason = `capability not permitted for sub-agent: ${call.name}`
      yield { type: "tool_denied", callId: call.id, toolName: call.name, reason } as ToolDeniedEvent
      yield {
        type: "tool_result",
        callId: call.id,
        name: call.name,
        content: reason,
        isError: true,
        errorKind: "governance_denied",
      } as ToolResultEvent
    }
    if (permitted.length > 0) {
      yield* this.inner.executeAll(permitted, ctx)
    }
  }
}

File: python/deepstrike/runtime/filtered_plane.py
from __future__ import annotations

from collections.abc import AsyncIterator

from deepstrike._kernel import ToolCall
from deepstrike.providers.stream import StreamEvent, ToolDeniedEvent, ToolResultEvent
from deepstrike.runtime.execution_plane import ExecutionPlane, RunContext

_DEFAULT_META = frozenset({"skill", "memory", "knowledge", "update_plan"})


class FilteredExecutionPlane(ExecutionPlane):
  """Wraps an execution plane, allowing only manifest-permitted tool IDs (+ meta-tools)."""

  def __init__(
    self,
    inner: ExecutionPlane,
    permitted_ids: set[str],
    meta_tools: frozenset[str] = _DEFAULT_META,
  ) -> None:
    self._inner = inner
    self._permitted = permitted_ids
    self._meta = meta_tools

  def register(self, *tools) -> "FilteredExecutionPlane":
    self._inner.register(*tools)
    return self

  def unregister(self, name: str) -> "FilteredExecutionPlane":
    self._inner.unregister(name)
    return self

  def schemas(self) -> list[dict]:
    return [
      s for s in self._inner.schemas()
      if s.name in self._permitted or s.name in self._meta
    ]

  async def execute_all(self, calls: list[ToolCall], ctx: RunContext) -> AsyncIterator[StreamEvent]:
    permitted: list[ToolCall] = []
    for call in calls:
      if call.name in self._meta or call.name in self._permitted:
        permitted.append(call)
        continue
      reason = f"capability not permitted for sub-agent: {call.name}"
      yield ToolDeniedEvent(call_id=call.id, tool_name=call.name, reason=reason)
      yield ToolResultEvent(
        call_id=call.id,
        name=call.name,
        content=reason,
        is_error=True,
        error_kind="governance_denied",
      )
    if permitted:
      async for evt in self._inner.execute_all(permitted, ctx):
        yield evt
